### PR TITLE
[JSC] Fix Atomics.wait by handling termination cases

### PR DESCRIPTION
**Pull request by hyjorc1**
#### dea928b186e9bd4da26e7bb09a7c6e8e013ac5e4
<pre>
[JSC] Fix Atomics.wait by handling termination cases
<a href="https://bugs.webkit.org/show_bug.cgi?id=278866">https://bugs.webkit.org/show_bug.cgi?id=278866</a>
<a href="https://rdar.apple.com/129234981">rdar://129234981</a>

Reviewed by Yusuke Suzuki.

Atomics.wait is used in worker threads. It should handle
termination cases; otherwise, a memory leak can occur when
the worker threads are waiting with an infinite timeout
while the threads are signaled for termination.

* Source/JavaScriptCore/runtime/AtomicsObject.cpp:
(JSC::atomicsWaitImpl):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::hasSyncWaiter):
(JSC::VM::syncWaiter):
* Source/JavaScriptCore/runtime/VM.h:
* Source/JavaScriptCore/runtime/VMTraps.cpp:
(JSC::VMTraps::fireTrap):
* Source/JavaScriptCore/runtime/WaiterListManager.cpp:
(JSC::WaiterListManager::waitSyncImpl):
(JSC::WaiterListManager::notifyWaiterImpl):
(JSC::WaiterListManager::unregister):
* Source/JavaScriptCore/runtime/WaiterListManager.h:

Canonical link: <a href="https://commits.webkit.org/282974@main">https://commits.webkit.org/282974@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/9cdb4eab9c59d04a72ff2588c2b7c94518b6b0d5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64691 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44056 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17287 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68714 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15299 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51836 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15576 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52015 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10554 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67757 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40763 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55972 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32637 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37429 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13366 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14170 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/57800 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59354 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13677 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70421 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/63933 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8636 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13174 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59346 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8670 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56057 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7134 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/820 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/85701 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9827 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39868 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15113 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40945 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42128 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40689 "Built successfully") | | |

### CODE REVIEW

File: Source/JavaScriptCore/runtime/VM.cpp [hunk 1]
```diff
@@ -1483,10 +1483,15 @@ bool VM::isScratchBuffer(void* ptr)
...
+{
+    return !!m_syncWaiter;
+}
 Ref<Waiter> VM::syncWaiter()
 {
-    m_syncWaiter->setVM(this);
-    return m_syncWaiter;
+    if (!m_syncWaiter)
+        m_syncWaiter = adoptRef(new Waiter(this));
+    return *m_syncWaiter;
```
Comment (Constellation) [inline Source/JavaScriptCore/runtime/VM.cpp]: I think it is fine to always have syncWaiter.

File: Source/JavaScriptCore/runtime/WaiterListManager.cpp [hunk 2]
```diff
@@ -224,11 +222,6 @@ void WaiterListManager::notifyWaiterImpl(const AbstractLocker& listLocker, Ref<W
         });
         return;
     }
-
-    // If waiter is a SyncWaiter, we null out its vm to indicate that this waiter
-    // is removed from the WaiterList.
-    ASSERT(waiter->vm());
-    waiter->clearVM(listLocker);
```
Comment (Constellation) [inline Source/JavaScriptCore/runtime/WaiterListManager.cpp]: what is the reason of this change?

File: Source/JavaScriptCore/runtime/VMTraps.cpp [hunk 3]
```diff
@@ -362,6 +363,11 @@ void VMTraps::fireTrap(VMTraps::Event event)
         m_condition->notifyAll(locker);
     }
 #endif
+
+    if (event == NeedTermination) {
+        if (vm().hasSyncWaiter())
+            vm().syncWaiter()->condition().notifyOne();
+    }
```
Comment (Constellation) [inline Source/JavaScriptCore/runtime/VMTraps.cpp]: Instead of doing this here, let's do it inside VMTraps::SignalSender::work, which can do this repeatedly until VM gets terminated. And also, we should do it after `sendMessage` gets sent. So you can just do it before the shutting down check.

File: Source/JavaScriptCore/runtime/WaiterListManager.cpp (hunk 2 above)
Comment (hyjorc1) [reply Source/JavaScriptCore/runtime/WaiterListManager.cpp]: Previously, we used whether `waiter->vm() == nullptr` as the indication of whether the waiter node is dequeued from the list or not. However, we can just use `waiter->isOnList()` since `waiter` has type `BasicRawSentinelNode`. See

```
bool didGetDequeued = !syncWaiter->isOnList();
```

in the above changes.

File: Source/JavaScriptCore/runtime/VMTraps.cpp (hunk 3 above)
Comment (hyjorc1) [reply Source/JavaScriptCore/runtime/VMTraps.cpp]: Discussed. Add the notification invocation to both here and `sendMessage`.

File: Source/JavaScriptCore/runtime/VM.cpp (hunk 1 above)
Comment (hyjorc1) [reply Source/JavaScriptCore/runtime/VM.cpp]: Fixed.

File: Source/JavaScriptCore/runtime/VM.cpp [hunk 4]
```diff
@@ -1485,7 +1485,6 @@ bool VM::isScratchBuffer(void* ptr)
 
 Ref<Waiter> VM::syncWaiter()
 {
-    m_syncWaiter->setVM(this);
```
Comment (Constellation) [inline Source/JavaScriptCore/runtime/VM.cpp]: Why do we need this?

File: Source/JavaScriptCore/runtime/WaiterListManager.cpp (hunk 2 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/WaiterListManager.cpp]: So, `syncWaiter->vm()` check in the while loop is not correct, right?

Comment (Constellation) [reply Source/JavaScriptCore/runtime/WaiterListManager.cpp]: I think given that this patch's intention is not changing that part, using the current mechanism is better.

File: Source/JavaScriptCore/runtime/VM.cpp (hunk 4 above)
Comment (hyjorc1) [reply Source/JavaScriptCore/runtime/VM.cpp]: Previously, we may need to `clearVM` for checking whether the node `isOnList`. Now, we don't need that since we never clear the vm any more.

**Comment by hyjorc1** [comment]
> So, syncWaiter->vm() check in the while loop is not correct, right?

Yes, we don't need that. Thanks for pointing it out.

**Comment by hyjorc1** [comment]
Ok, let's split the patch. And this one should focus on fixing the potential memory leak for waitSync.

**Review by Constellation — APPROVED** [review]
r=me
